### PR TITLE
[BE] Enable ruff's UP rules and autoformat nn/ mps/ and torch/

### PR DESCRIPTION
**Pull request by justinchuby**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #105538
* #105437
* __->__ #105436

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

### CODE REVIEW

File: torch/__init__.py [hunk 1]
```diff
@@ -16,8 +16,6 @@
 import textwrap
 import ctypes
 import inspect
-if sys.version_info < (3,):
-    raise Exception("Python 2 has reached end-of-life and is no longer supported by PyTorch.")
```
Comment (justinchuby) [inline torch/__init__.py:20]: Revert

Comment (malfet) [reply torch/__init__.py:20]: actually, this can go away...

File: torch/_torch_docs.py [hunk 2]
```diff
@@ -1,4 +1,3 @@
-# -*- coding: utf-8 -*-
```
Comment (malfet) [inline torch/_torch_docs.py:1]: Why are you removing `coding` decorator for the file that contains non-ascii characters?
```python
>>> open("torch/_torch_docs.py",encoding="ascii").read()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/malfet/miniconda3/lib/python3.10/encodings/ascii.py", line 26, in decode
    return codecs.ascii_decode(input, self.errors)[0]
UnicodeDecodeError: 'ascii' codec can't decode byte 0xe2 in position 66808: ordinal not in range(128)
```

File: torch/serialization.py [hunk 3]
```diff
@@ -1085,11 +1085,11 @@ def _check_container_source(container_type, source_file, original_source):
                         if file_size == 0:
                             f.write(lines)
                         elif file_size != len(lines) or f.read() != lines:
-                            raise IOError
+                            raise OSError
```
Comment (malfet) [inline torch/serialization.py:1088]: Can you explain the rationale behind replacing IOError with OSError?

File: torch/nn/modules/module.py [hunk 4]
```diff
@@ -2140,11 +2140,11 @@ def load(module, local_state_dict, prefix=''):
...
-                        ', '.join('"{}"'.format(k) for k in unexpected_keys)))
+                        ', '.join(f'"{k}"' for k in unexpected_keys)))
             if len(missing_keys) > 0:
                 error_msgs.insert(
                     0, 'Missing key(s) in state_dict: {}. '.format(
-                        ', '.join('"{}"'.format(k) for k in missing_keys)))
+                        ', '.join(f'"{k}"' for k in missing_keys)))
 
         if len(error_msgs) > 0:
             raise RuntimeError('Error(s) in loading state_dict for {}:\n\t{}'.format(
```
Comment (malfet) [inline torch/nn/modules/module.py:2150]: Err, why don't you replace this one?

**Review by malfet — APPROVED** [review]

File: torch/serialization.py (hunk 3 above)
Comment (justinchuby) [reply torch/serialization.py:1088]: It's an autofix. According to https://beta.ruff.rs/docs/rules/os-error-alias/, IOError is an alias of OSError

File: torch/nn/modules/module.py (hunk 4 above)
Comment (justinchuby) [reply torch/nn/modules/module.py:2150]: It's all autofix by ruff. I can do a manual pass for this one!

**Review by albanD — APPROVED** [review]
Sounds good

File: torch/nn/modules/module.py (hunk 4 above)
Comment (justinchuby) [reply torch/nn/modules/module.py:2150]: #105538

**Comment by justinchuby** [comment]
@albanD @malfet The dynamo tests raises the following (somehow the maxpool2d test was routed to use maxpool1d). Is this related to f-string handling in dynamo? I tried but could not reproduce locally on an m1 mac.

I used

```
TEST_WITH_TORCHDYNAMO=1 python test/nn/test_pooling.py -k test_MaxPool2d_indices_cpu_float32
```

The error:

```
________ TestPoolingNNDeviceTypeCPU.test_MaxPool2d_indices_cpu_float32 _________
Traceback (most recent call last):
  File "/var/lib/jenkins/workspace/test/nn/test_pooling.py", line 1269, in test_MaxPool2d_indices
    self._test_maxpool_indices(2, device=device, dtype=dtype)
  File "/var/lib/jenkins/workspace/test/nn/test_pooling.py", line 1222, in _test_maxpool_indices
    output, indices = module(input_var)
  File "/opt/conda/envs/py_3.11/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1522, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/envs/py_3.11/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1531, in _call_impl
    return forward_call(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/envs/py_3.11/lib/python3.11/site-packages/torch/nn/modules/pooling.py", line 92, in forward
    return F.max_pool1d(input, self.kernel_size, self.stride,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/envs/py_3.11/lib/python3.11/site-packages/torch/_jit_internal.py", line 486, in fn
    return if_true(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/envs/py_3.11/lib/python3.11/site-packages/torch/nn/functional.py", line 680, in max_pool1d_with_indices
    return torch.max_pool1d_with_indices(input, kernel_size, stride, padding, dilation, ceil_mode)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
RuntimeError: Expected 2 to 3 dimensions, but got 4-dimensional tensor for argument #1 'self' (while checking arguments for max_pool1d)
```

https://github.com/pytorch/pytorch/actions/runs/5604098622/jobs/10251876391?pr=105436#step:15:1232

File: test/nn/test_pooling.py [hunk 5]
```diff
@@ -1209,9 +1209,9 @@ def expected_output(dim, dtype):
                 return torch.stack([col, col + 2], 1).view(2, 2, 2, 2)
 
         if adaptive:
-            cls_name = 'AdaptiveMaxPool{}d'.format(num_dim)
+            cls_name = f'AdaptiveMaxPool{num_dim}d'
         else:
-            cls_name = 'MaxPool{}d'.format(num_dim)
+            cls_name = f'MaxPool{num_dim}d'
```
Comment (albanD) [inline test/nn/test_pooling.py:1214]: Could this change in formatting make the formatting behavior differently?

Comment (justinchuby) [reply test/nn/test_pooling.py:1214]: So I tested on linux and could not reproduce either. I am going to revert this line and see what happens on CI

Comment (justinchuby) [reply test/nn/test_pooling.py:1214]: So it doesn't fail anymore. I can't think of a reason why it's happening. Created https://github.com/pytorch/pytorch/issues/105716 to track

**Pull request closed without merging**